### ISSUE TITLE
Advanced Sorting / Filtering

### ISSUE THREAD

**Issue by mehmetbajin** [OMSCentral/omscentral#3]
In any review card list view...

- ...allow user to filter by attributes such as Semester, Difficulty, Workload, Rating, etc.
- ...allow user to sort by appropriate attributes also.

**Comment by SohamGhormade** [comment]
As of May 29, 2021, filtering and sorting by semester is already supported which is good.
Snapshot for reference:
![image](https://user-images.githubusercontent.com/9662026/120086416-b2b44a80-c0ac-11eb-867c-69db7525926f.png)

Given that observation, the remaining work for this enhancement would be to provide filtering and sorting functionality for available attributes : 
- Avg. Difficulty
- Workload
- Rating  
## Notes
### Development
- Keying off Semester for the rest can be a good start.
- Heads up: unlike Semester which is discrete, other attributes are continuous e.g decimals between 1-5 for Rating. They are internally mapped to Medium, Hard, etc
 
### Testing
- Finding out how to add 5 reviews to local database for testing is a good next step.
- Then test out functionality for each of the attributes.
- Potential bonus : Integration testing  for this functionality.

Planning to looking into this issue.
Will assign myself once I have a basic prototype working for Difficulty.

**Comment by mehmetbajin** [comment]
Thank you so much!

**Comment by SohamGhormade** [comment]
Mehmet, one related question, how to add a review when testing locally?

Asking because when I attempt to register myself as a user to add a review, I see "Unhandled Rejection (Error): Forbidden"

**Comment by mehmetbajin** [comment]
Did you configure authentication in your Firebase project?

**Comment by SohamGhormade** [comment]
Hi Mehmet,
As an update, from my end, I setup authenication as advised.
Specifically, I have enabled Email/Password and Google authentication as seen in the snapshot below:
![image](https://user-images.githubusercontent.com/9662026/120234822-bb309080-c226-11eb-950f-8e0029af1dfb.png)

With authentication enabled, in terms of progress, I am able to login which is good.
But when I write and submit a review I get a 403 error code with message "Operation Forbidden".

Snapshot for reference:
![image](https://user-images.githubusercontent.com/9662026/120234923-f4690080-c226-11eb-94b0-53b9c0b30259.png)

The following gets logged on the terminal window:
```
[server] 2021-05-31T19:37:55.906Z - POST: [10218] 200 /graphql (17.471 ms)
[server] error: Forbidden {"data":null,"isBoom":true,"isServer":false,"output":{"statusCode":403,"payload":{"statusCode":403,"error":"Forbidden","message":"Forbidden"},"headers":{}},"stack":"Error: Forbidden\n    at Object.fallbackError (/Users/sohamghormade/Documents/Georgia Tech/Code/omscentral/server/src/api/permissions.ts:42:15)\n    at middleware (/Users/sohamghormade/Documents/Georgia Tech/Code/omscentral/server/node_modules/graphql-shield/src/generator.ts:73:32)\n    at processTicksAndRejections (node:internal/process/task_queues:94:5)","timestamp":"2021-05-31 15:38:07"}
[server] 2021-05-31T19:38:07.329Z - POST: [10218] 200 /graphql (5.614 ms)
``` 
Basically, it seems like I cannot write a review, even after logging in, in my local env, due to a permissions issue...

**Comment by mehmetbajin** [comment]
This is probably caused by line 11 of `server/src/api/mutation/insertReview.ts`.

What is the output of the following in your postgres database?

```
select * from omscentral_user;
```

**Comment by SohamGhormade** [comment]
The table for omscentral_user is empty per the output below:
```
postgres=# \connect omscentral_scratch 
You are now connected to database "omscentral_scratch" as user "sohamghormade".
omscentral_scratch=# \dt
                    List of relations
 Schema |           Name            | Type  |   Owner    
--------+---------------------------+-------+------------
 public | omscentral_config         | table | omscentral
 public | omscentral_course         | table | omscentral
 public | omscentral_course_metric  | table | omscentral
 public | omscentral_migration      | table | omscentral
 public | omscentral_migration_lock | table | omscentral
 public | omscentral_program        | table | omscentral
 public | omscentral_review         | table | omscentral
 public | omscentral_semester       | table | omscentral
 public | omscentral_session        | table | omscentral
 public | omscentral_specialization | table | omscentral
 public | omscentral_user           | table | omscentral
(11 rows)

omscentral_scratch=# select * from omscentral_user 
omscentral_scratch-# 
```

**Comment by mehmetbajin** [comment]
When the application is running on your local development environment (upon `npm run dev`) and you register for an account using email/password, the application inserts a new record into `omscentral_user` table.

In order for a user to insert a review record, that user's record must first exist in the `omcentral_user` table. In addition, the author for the inserted review must be the currently authenticated user.

Since there are no user records in the `omscentral_user` table, HTTP 403 error blocks creating a review.

I recommend doing the following to remediate:

1. Register for a user account using `http://localhost:3000/register` (if the account already exists in Firebase, delete the account from Firebase Authentication).

Once registered, you'll be automatically signed in and should be able to start composing reviews. You can verify registration succeeded by observing the insertion of a user record to `omscentral_user` table (in addition to creation of the user account in Firebase Authentication).

**Comment by SohamGhormade** [comment]
Mehmet, thanks for sharing the relevant background information in context of the 403 error I am seeing.
it makes sense to me.

From my end, when I register a new user, what I am seeing, however, is that this request does not get to Firebase for authentication.
I mean I see a green check in the webpage but I also see 
- 403 on the same webpage
- No users on Firebase
Attached a mp4 video for context:

https://user-images.githubusercontent.com/9662026/120240950-d227b080-c22f-11eb-8008-57ab3d0ae867.mp4

Basically, I am able to sign in as a new user but the request does not get to Firebase and also not to the database as seen in the select * for the omscentral_user table in the video above.

One thing worth noting, from my end, is that I did the autentication for Firebase after I created it.
This is likely okay because I am seeing analytics data on Firebase i.e I see a user in my location i.e me attempting to access the website.

So I am guessing authentication setup is okay. If not, I can create a new Firebase DB with authentication setup if you recommend that this approach will help.

**Comment by mehmetbajin** [comment]
Thanks for uploading the video.

Can you do the following?

1. Update `client/src/config/firebase.ts` to log the config to the browser console by appending the `console.log('firebase.config:', config);` at line 23
2. Start the application in local development mode
3. Look at the browser console once the application starts up on `http://localhost:3000`

Can you send a screenshot of what you see in the browser console after steps 1-3?

**Comment by SohamGhormade** [comment]
Following steps 1-3, 
I see the following exception (400 error code) in the Chrome developer console:
```
index.js:1 FirebaseError: Installations: Create Installation request failed with error "400 INVALID_ARGUMENT: 
Request contains an invalid argument." (installations/request-failed).
    at https://localhost:3000/static/js/vendors~main.chunk.js:25613:27
    at step (https://localhost:3000/static/js/vendors~main.chunk.js:27587:17)
    at Object.next (https://localhost:3000/static/js/vendors~main.chunk.js:27518:14)
    at fulfilled (https://localhost:3000/static/js/vendors~main.chunk.js:27473:24)
console.<computed> @ index.js:1
Node cannot be found in the current page.
``` 
Screenshot for reference, generated after step 3:
![image](https://user-images.githubusercontent.com/9662026/120258863-d0c2ac00-c260-11eb-9aa2-27e86fa663a8.png)

**Comment by SohamGhormade** [comment]
As an update, Mehmet, your comment [here](https://github.com/OMSCentral/omscentral/issues/137#issuecomment-850904252) fixed the firebase issue.
Basically, REACT_APP_FIREBASE_AUTH_DOMAIN and REACT_APP_FIREBASE_API_KEY from staging were used for client/env.development.local which was not right.
Simply put, double checking that client side local env vars are set correctly helped fix the issue after looking at the comments about a similar issue [here](https://github.com/angular/angularfire/issues/2546)

Overall, as an update, I am able to login with different user names and publish reviews which is good from testing POV.
## Snapshot:
![image](https://user-images.githubusercontent.com/9662026/120417981-c0c4ce00-c32d-11eb-8568-bffd608fed4b.png)

## Next step
-Will try out adding filter support for difficulty and verify change

## Note
- I am seeing the reviews persist on rerunning npm run dev. I also see the users I created show in Firebase as expected.
- This is good. However, the postgres DB table for omscentral_user is empty. This part is odd.

**Comment by mehmetbajin** [comment]
> This is good. However, the postgres DB table for omscentral_user is empty. This part is odd.

How are you verifying this? When verifying, are you connected to the same postgres DB as the api?

**Comment by SohamGhormade** [comment]
I created only one DB called omscentral-scratch using postgres.
I am connecting to it.
So I am expecting it to be the right one.
In case there is a way to verify, please share.
Thanks!

postgres sql command line output for reference, if needed:
```
(base) sohamghormade@Sohams-MacBook-Air omscentral % psql postgres
psql (13.3)
Type "help" for help.
postgres=# \l
                                         List of databases
        Name        |     Owner     | Encoding | Collate | Ctype |        Access privileges        
--------------------+---------------+----------+---------+-------+---------------------------------
 omscentral_scratch | sohamghormade | UTF8     | C       | C     | 
 postgres           | sohamghormade | UTF8     | C       | C     | 
 template0          | sohamghormade | UTF8     | C       | C     | =c/sohamghormade               +
                    |               |          |         |       | sohamghormade=CTc/sohamghormade
 template1          | sohamghormade | UTF8     | C       | C     | =c/sohamghormade               +
                    |               |          |         |       | sohamghormade=CTc/sohamghormade
(4 rows)

postgres=# 
postgres=# 
postgres=# \connect omscentral_scratch 
You are now connected to database "omscentral_scratch" as user "sohamghormade".
omscentral_scratch=# \dt
                    List of relations
 Schema |           Name            | Type  |   Owner    
--------+---------------------------+-------+------------
 public | omscentral_config         | table | omscentral
 public | omscentral_course         | table | omscentral
 public | omscentral_course_metric  | table | omscentral
 public | omscentral_migration      | table | omscentral
 public | omscentral_migration_lock | table | omscentral
 public | omscentral_program        | table | omscentral
 public | omscentral_review         | table | omscentral
 public | omscentral_semester       | table | omscentral
 public | omscentral_session        | table | omscentral
 public | omscentral_specialization | table | omscentral
 public | omscentral_user           | table | omscentral
(11 rows)
omscentral_scratch=# 
omscentral_scratch=# select * from omscentral_user 
omscentral_scratch-# select * from omscentral_migration  // no data 
omscentral_scratch-# 
```
Basically, finding and connecting to DB and then accessing table using SELECT * as before.

**Comment by mehmetbajin** [comment]
Try putting a semi-colon at the end of `select * from omscentral_user;` and `select * from omscentral_migration;`.

How does that change the output, if at all?

**Comment by SohamGhormade** [comment]
Putting the semi colon was key.
With semi colon put in, I am seeing output which is good.
Will keep in mind, for future reference that sql commands require semi-colon at the end , thanks!
```
omscentral_scratch=# select * from omscentral_user;
              id              | auth_provider | password_hash | password_salt |     email     | name | role  | photo_url | program_id | specialization_id | anonymous | last_signed_in |    created    |    updated    
------------------------------+---------------+---------------+---------------+---------------+------+-------+-----------+------------+-------------------+-----------+----------------+---------------+---------------
 tjohyPuVMmTIPOu17RTUE3MZyxa2 | password      |               |               | r@r.com       | r    | basic |           |            |                   | t         |  1622602534820 | 1622602535081 |              
 V1wBG8t4SgbzWzD0l7rXLXhWPXx2 | password      |               |               | blah@blah.com | blah | basic |           |            |                   | t         |  1622603138656 | 1622602722234 | 1622603138932
```
In terms of next step, will look into filtering by difficulty as planned before.

**Comment by SohamGhormade** [comment]
As an update, I have a draft PR for partially addressing this  issue.
Specifically, the draft PR adds support for filtering by difficulty.

The draft PR is  #142.
The PR is not ready for review yet.
That is because I have some clean up to do as noted in the comments section of the PR.
Once the PR is ready, it can be reviewed by Mehmet.

**Comment by SohamGhormade** [comment]
As an update, staging branch has the full end to end change to add filtering and sorting functionality for difficulty attribute.
(Relevant PRs : #142 with filtering change and #145 with with commit with sorting related change)

In terms of next steps, the same pattern, as the one used for difficulty, can be used for other attributes e.g rating and workload.

From my end, in terms of next steps, I am planning to look at issue #6 next.
However, others, e.g new contributors can look into this enhancement or I can pick it up once other issues are resolved
